crates/datasources/src/lib.rs

```diff
@@ -17,3 +30,49 @@ pub mod postgres;
 pub mod snowflake;
 pub mod sqlite;
 pub mod sqlserver;
+
+
+pub type DatasourceError = Box<dyn std::error::Error + Send + Sync>;
+#[async_trait]
+pub trait Datasource: Send + Sync {
+    fn name(&self) -> &'static str;
+
+    /// Validate the provided tunnel options.
+    fn validate_tunnel_connections(
+        &self,
+        tunnel_opts: Option<&TunnelOptions>,
+    ) -> Result<(), DatasourceError>;
+
+    /// Validate the provided credentials.
+    fn validate_credentials(
+        &self,
+        creds: Option<CredentialsOptions>,
+    ) -> Result<(), DatasourceError>;
```

Comment:
missed this comment in my previous review: 

I feel like there should be default implementations of this so that if there aren't any credentials or options then implementors shouldn't have to implement? 

Also:
- confused why `TunnelOptions` is a ref but credentials isn't.
- not sure that we need to have a reference to the object?

Reply:
I think being explicit makes sense, but I think that the default implementations could/should error so that if you _did_ specify tunnel for a file-based protocol, you'd get an error, (which is correct,) and implementers of datasources wouldn't have to think if tunnels made sense for their datasource?

Reply:
I could see an argument for the inverse as you suggested. (the tunnel having control over the datasources).

something like

```rs
pub trait Tunnel {
  fn validate_datasource(&dyn Datasource) -> Result<()> 
}
```

Unfortunately, our current architecture tightly couples the datasources to the tunnels, so i think this would be a larger change outside the scope of the table options refactor.

Reply:
I would have thought that this refactor could shim a different interface on top of our current architecture.

Regardless we should leave comments here to describe the desired endstate, and make sure we track getting there.

Reply:
I just removed these validate methods for now. I think having the tunnel/credentials validate the datasource makes a lot more sense & on further inspection that is what we are already doing. So i just removed those and we can continue to use the current logic.